### PR TITLE
better http server init; move init/refresh out of components; add audio support

### PR DESCRIPTION
**Pull request by songgao**
also use mimeType from http server in the popup. More mimeType migration
in future PRs.

### CODE REVIEW

**Comment by songgao** [comment]
Actually please hold off for review. I'm gonna make a few more changes.

**Comment by songgao** [comment]
OK; ready for review again! The android part still needs some more work, especially now that chat is using it too. I'm tracking that part in KBFS-2974, and eventually we can remove `react-native-mime-types`.

File: shared/fs/filepreview/default-view.js [hunk 1]
```diff
@@ -55,17 +55,7 @@ const DefaultView = (props: DefaultViewProps) => (
         onClick={props.onShare}
       />
     )}
-    {isIOS ? (
```
Comment (jzila) [inline shared/fs/filepreview/default-view.js]: Why is this removed?

File: shared/fs/filepreview/pdf-view.native.js [hunk 2]
```diff
@@ -1,16 +1,15 @@
...
-
-const getOnMessage = onInvalidToken => e =>
-  e.nativeEvent.data.title === Constants.invalidTokenTitle && onInvalidToken()
-
 const PdfView = ({url, onInvalidToken}: PdfViewProps) => (
-  <WebView source={{uri: url}} injectedJavaScript={webviewJS} onMessage={getOnMessage(onInvalidToken)} />
+  <WebView styles={stylesVideoView} source={{uri: url}} />
 )
 
+const stylesVideoView = {
```
Comment (jzila) [inline shared/fs/filepreview/pdf-view.native.js]: Shouldn't this be named `stylesPdfView`?

File: shared/fs/popups/row-action-popup-container.js [hunk 3]
```diff
@@ -159,14 +167,24 @@ const mergeProps = (stateProps, dispatchProps) => {
...
   }
 }
 
-export default compose(connect(mapStateToProps, mapDispatchToProps, mergeProps), setDisplayName('Popup'))(
-  Popup
-)
+export default compose(
+  connect(mapStateToProps, mapDispatchToProps, mergeProps),
+  lifecycle({
+    componentDidMount() {
```
Comment (jzila) [inline shared/fs/popups/row-action-popup-container.js]: Is there a risk of a race here? I just want to make sure that we've loaded the mime type before rendering the popup.

File: shared/fs/filepreview/av-view.native.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+// @flow
+import * as React from 'react'
+import {globalColors} from '../../styles'
+import WebView from '../../common-adapters/web-view.native'
```
Comment (jzila) [inline shared/fs/filepreview/av-view.native.js]: Is there a reason not to have a single adapter for both desktop and native webviews? They seem to have the same API, so we might as well make them the same. Would avoid having to have this `.native.js` file.

File: shared/fs/filepreview/default-view.js (hunk 1 above)
Comment (songgao) [reply shared/fs/filepreview/default-view.js]: There's no way this code gets reached. This is in the "default view", which is only used when we don't have a viewing component. For media types, we just display them in their dedicated components.

File: shared/fs/filepreview/av-view.native.js (hunk 4 above)
Comment (songgao) [reply shared/fs/filepreview/av-view.native.js]: They actually have different APIs. They accept different forms of parameters to specify source URI; they also use different interface for injecting JavaScript/CSS. But I guess we can still unify them by shipping our own Props. Will do; thanks!

File: shared/fs/popups/row-action-popup-container.js (hunk 3 above)
Comment (songgao) [reply shared/fs/popups/row-action-popup-container.js]: Yes, but it's no harm. In fact, the first time you see this popup on an image, it won't show a "Save" button initially, and the Save button pops up instantly. We could probably move this up to before `routeAppend` if this turns out to be annoying.

File: shared/common-adapters/web-view.js.flow [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+import * as React from 'react'
+
+export type WebViewInjections = {
+  javaScript?: string,
+  css?: string,
+}
+
+export type WebViewProps = {
+  url: string,
+	injections?: WebViewInjections,
```
Comment (jzila) [inline shared/common-adapters/web-view.js.flow]: whitespace

Comment (songgao) [reply shared/common-adapters/web-view.js.flow]: Ah; Looks like eslint doesn't go through .flow files

File: shared/actions/fs/index.js [hunk 6]
```diff
@@ -282,6 +282,56 @@ function* ignoreFavoriteSaga(action: FsGen.FavoriteIgnorePayload): Saga.SagaGene
   }
 }
 
+const getMimeTypePromise = (path: Types.Path, serverInfo: Types._LocalHTTPServer) =>
+  new Promise((resolve, reject) =>
+    getMimeTypeFromURL(Constants.generateFileURL(path, serverInfo), ({error, statusCode, mimeType}) => {
+      if (error !== undefined) {
```
Comment (akalin-keybase) [inline shared/actions/fs/index.js]:
```
if (error) {
  reject(error)
}
```
is more idiomatic i think

File: shared/actions/fs/index.js [hunk 7]
```diff
@@ -305,20 +355,46 @@ function* fileActionPopup(action: FsGen.FileActionPopupPayload): Saga.SagaGenera
...
-      path: action.payload.path,
-      mimeType,
-    })
+  let bare = false
+  if (pathItem.type === 'file') {
+    let mimeType = pathItem.mimeType
+    if (mimeType === '') {
+      mimeType = yield Saga.call(_loadMimeType, path)
+    }
+    bare = isMobile && ['image'].includes(Constants.viewTypeFromMimeType(mimeType))
```
Comment (akalin-keybase) [inline shared/actions/fs/index.js]: `Constants.viewTypeFromMimeType(mimeType) === 'image'`?

File: shared/actions/fs/index.js [hunk 8]
```diff
@@ -332,7 +408,7 @@ function* fsSaga(): Saga.SagaGenerator<any, any> {
   yield Saga.safeTakeEvery(FsGen.filePreviewLoad, filePreview)
   yield Saga.safeTakeEvery(FsGen.favoritesLoad, listFavoritesSaga)
   yield Saga.safeTakeEvery(FsGen.favoriteIgnore, ignoreFavoriteSaga)
-  yield Saga.safeTakeEveryPure(FsGen.mimeTypeLoad, loadMimeType, loadMimeTypeResult)
+  yield Saga.safeTakeEvery(FsGen.mimeTypeLoad, loadMimeType)
```
Comment (akalin-keybase) [inline shared/actions/fs/index.js]: this can still be pure I think, just have `loadMimeType` return whatever it's yielding

File: shared/common-adapters/web-view.desktop.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+// @flow
+import * as React from 'react'
+import type {WebViewProps} from './web-view'
+
+class WebView extends React.PureComponent<WebViewProps> {
+  webviewRef: any
```
Comment (akalin-keybase) [inline shared/common-adapters/web-view.desktop.js]: better type? see https://github.com/facebook/flow/pull/5920

File: shared/common-adapters/web-view.desktop.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+  constructor(props: WebViewProps) {
+    super(props)
+    this.webviewRef = React.createRef()
+  }
+  componentDidMount() {
+    const css = (this.props.injections && this.props.injections.css) || ''
+    const javaScript = (this.props.injections && this.props.injections.javaScript) || ''
+    if (css || javaScript) {
+      this.webviewRef.current.addEventListener('dom-ready', () => {
```
Comment (akalin-keybase) [inline shared/common-adapters/web-view.desktop.js]: pretty sure you'll need to test `this.webviewRef.current`, especially if you type `webviewRef`

File: shared/common-adapters/web-view.js.flow [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+  css?: string,
+}
+
+export type WebViewProps = {
+  url: string,
+  injections?: WebViewInjections,
+  style?: Object,
+}
+
+declare export default React.ComponentType<WebViewProps>
```
Comment (akalin-keybase) [inline shared/common-adapters/web-view.js.flow]: This component makes me a bit nervous...can you explain a bit if there are any risks with this?

File: shared/fs/popups/row-action-popup-container.js [hunk 12]
```diff
@@ -159,14 +167,24 @@ const mergeProps = (stateProps, dispatchProps) => {
...
+      if (this.props.needLoadMimeType) {
+        this.props.loadMimeType()
+      }
+    },
+  }),
+  setDisplayName('Popup')
```
Comment (akalin-keybase) [inline shared/fs/popups/row-action-popup-container.js]: probably want to move this `setDisplayName` up

Comment (songgao) [reply shared/fs/popups/row-action-popup-container.js]: Ah right! I forgot to fix this.

File: shared/common-adapters/web-view.js.flow (hunk 11 above)
Comment (songgao) [reply shared/common-adapters/web-view.js.flow]: We are filtering returned mimeType in the KBFS http server to never return text/html or text/javascript, so there's that. We've also disabled download in electron. I think if we limit the use of this to internal URLs only it's probably fine. We also have the option to use segregated session for these webviews, which I haven't done but think is possible.

Are there specific risks you are thinking of?

File: shared/common-adapters/web-view.native.js [hunk 13]
```diff
@@ -1,8 +1,38 @@
...
 
-export default wv
+const combineJavaScriptAndCSS = (injections?: WebViewInjections) =>
+  !injections
+    ? ''
+    : `
+(function() {
+  const node = document.createElement('style')
+  document.body.appendChild(node)
+  node.innerHTML = \` ${sanitize(injections.css)} \`
```
Comment (akalin-keybase) [inline shared/common-adapters/web-view.native.js]: setting `innerHTML` gives me the heebie-jeebies. Try
```
const style = document.createElement('style');
document.head.appendChild(style);
style.type = 'text/css';
style.appendChild(document.createTextNode(css));
```

from https://stackoverflow.com/questions/524696/how-to-create-a-style-tag-with-javascript

File: shared/common-adapters/web-view.native.js [hunk 14]
```diff
@@ -1,8 +1,38 @@
 // @flow
+import * as React from 'react'
 import {isIOS} from '../constants/platform'
 import {WebView} from 'react-native'
 import WKWebView from 'react-native-wkwebview-reborn'
+import type {WebViewInjections, WebViewProps} from './web-view'
+import {memoize} from 'lodash-es'
 
-const wv = isIOS ? WKWebView : WebView
+const sanitize = (str?: string): string => (str ? str.replace(/\\/g, '\\\\').replace(/`/g, '\\`') : '')
```
Comment (akalin-keybase) [inline shared/common-adapters/web-view.native.js]: `sanitize` seems like the wrong term, maybe `escape`?

File: shared/common-adapters/web-view.js.flow (hunk 11 above)
Comment (akalin-keybase) [reply shared/common-adapters/web-view.js.flow]: Ah, I made separate comments re. the injection of CSS. I took a closer look at how we interact with the http server and it seems ok, although I wonder if it would be clearer to pass keybase paths to `WebView` and have it do the path-to-URL conversion (but that can be in a separate PR)

File: shared/common-adapters/web-view.native.js (hunk 13 above)
Comment (songgao) [reply shared/common-adapters/web-view.native.js]: Sounds good! I didn't realize this option. Will do!

File: shared/common-adapters/web-view.js.flow (hunk 11 above)
Comment (songgao) [reply shared/common-adapters/web-view.js.flow]: Good idea. Will do!

Comment (songgao) [reply shared/common-adapters/web-view.js.flow]: @akalin-keybase I tried to have WebView do the URL converting, but it seems it's gonna be pretty complicated for some cases. For example the AVView uses a `about:blank` src, and use JavaScript to load the content dynamically. So either WebView should understand the video content, or it can't really control the URL being loaded. I guess having the WebView understand different types of content loaded is the right direction. What do you think?

(Either way it seems this is gonna have be a separate PR)

File: shared/common-adapters/web-view.desktop.js [hunk 15]
```diff
@@ -0,0 +1,32 @@
+// @flow
+import * as React from 'react'
+import type {WebViewProps} from './web-view'
+
+class WebView extends React.PureComponent<WebViewProps> {
+  _webviewRef: ?WebviewElement
+
+  _setWebviewRef = (r: ?HTMLElement) => {
+    this._webviewRef = (r: any)
```
Comment (songgao) [inline shared/common-adapters/web-view.desktop.js]: @akalin-keybase so I poked around and this is the best I could type it. Please let me know if you have better ideas!

Comment (akalin-keybase) [reply shared/common-adapters/web-view.desktop.js]: This can be
```
_setWebviewRef = (r: ?WebviewElement) => {
```
if you're in this area of the code again!

Comment (songgao) [reply shared/common-adapters/web-view.desktop.js]: I think I tried it, but flow didn't like me passing a `(r: ?WebviewElement): void` to `ref`. It expects ` `(r: ?HTMLElement): void`.
